src/assign.c

```diff
@@ -152,13 +152,24 @@ static SEXP shallow(SEXP dt, SEXP cols, R_len_t n)
   R_len_t i,l;
   int protecti=0;
   SEXP newdt = PROTECT(allocVector(VECSXP, n)); protecti++;   // to do, use growVector here?
-  //copyMostAttrib(dt, newdt);   // including class
-  DUPLICATE_ATTRIB(newdt, dt);
+  SET_ATTRIB(newdt, shallow_duplicate(ATTRIB(dt)));
+  SET_OBJECT(newdt, OBJECT(dt));
+  //SHALLOW_DUPLICATE_ATTRIB(newdt, dt);  // SHALLOW_DUPLICATE_ATTRIB would be a bit neater but is only available from R 3.3.0
+  
   // TO DO: keepattr() would be faster, but can't because shallow isn't merely a shallow copy. It
   //        also increases truelength. Perhaps make that distinction, then, and split out, but marked
   //        so that the next change knows to duplicate.
-  //        Does copyMostAttrib duplicate each attrib or does it point? It seems to point, hence DUPLICATE_ATTRIB
-  //        for now otherwise example(merge.data.table) fails (since attr(d4,"sorted") gets written by setnames).
+  //        keepattr() also merely points to the entire attrbutes list and thus doesn't allow replacing
+  //        some of its elements.
+  
+  // We copy all attributes that refer to column names so that calling setnames on either
+  // the original or the shallow copy doesn't break anything.
+  SEXP index = PROTECT(getAttrib(dt, sym_index)); protecti++;
```

Comment:
Protect AFAIK is not needed. `dt` would need to be garbage collected before we will assign index as an attribute of `newdt`.

Reply:
You are probably right but I wasn't quite sure so decided to follow the pattern of just a few lines later and take the cautious approach. Looking through the rest of the file, there are a few instances where the call to `getAttrib` is not `PROTECT`ed and many where it is. I also don't think that `dt` could be garbage collected.

Anyway, I wouldn't mind taking it out but I don't think it hurts and then it begs the question of what to do further down in the function or in the rest of the code.

Reply:
Agree, I spotted in the code there were protects sometimes where it was not needed. It is not really an issue, but it spreads the pattern of overprotecting, as you see on yourself. I used to do it as well. Still, I might be wrong, but this will be verified when preparing release for CRAN and running strict memory tests.

Reply:
I've removed the PROTECT statements and also went ahead and removed the one on what is now line 174 for consistency. I considered just going through the entire file while I was at it, but that seemed out of scope for the PR.

Reply:
I had the same thoughts and logic as you both. But `rchk` revealed that `getAttrib` can sometimes allocate inside it, perhaps more so now with ALTREP. Search for "A common source of true errors is a failure to protect the result of getAttrib when retrieving an attribute that may be automatically generated/converted (e.g. names, dimnames)" in https://github.com/kalibera/rchk/blob/master/doc/USAGE.md. Also search https://github.com/kalibera/rchk/blob/master/doc/INTERNALS.md for `getAttrib`.
I suspect in data.table's usage of the R API, we will never see `getAttrib` allocate, but R API is more general. So to pass `rchk` (as per steps in CRAN_Release.cmd, and as required by CRAN under additional checks) we have to protect getAttrib calls.
The variance in why some `getAttrib` calls are not protected in data.table, may be that `rchk` knows some cases of `getAttrib` do not need to be protected depending on what the 2nd argument is.
In the past I've gone through and protected any `getAttrib` calls that `rchk` spots until it passes.

Reply:
Very interesting, thanks for that. I don't mind reverting those changes at all. I just have two thoughts:
1) From reading the docs you linked, the seemingly inconsistent calls to `PROTECT` may be related to garbage collection only being triggered by allocating calls as not all instances of `getAttrib(*, R_NamesSymbol)` are currently protected.
2) As far as I can tell, the only instances where it does allocate are for rownames of a data.frame and with `R_NamesSymbol` when the first argument is a pairlist or a language object but then in the latter case it still calls `MARK_NOT_MUTABLE`, which I _thought_ protected the object from garbage collection.

Reply:
1. I'm not fully following there. But I did a grep and yes I see that you're right and not all `getAttrib(*, R_NamesSymbol)` are protected, for example in `rbindlist.c`. It could be that those will be picked up when I rerun `rchk` before release (perhaps those lines were added or changed in dev since 1.12.8),  or more likely, from prior discussions with Thomas I suspect that `rchk` looks afterwards to the usage of the unprotected `getAttrib`. If there is no possible GC between the getAttrib and its last usage, or if the result is protected by an assign using `setAttrib`, then `rchk` is clever enough (I guess) to not raise it. It does an extreme amount of tracking to spot unbalanced protection, for example, all statically (without runtime tests) so I've learnt not to underestimate how advanced it is.
2. That's one case.  But isn't it possible too that column names are also some kind of ALTREP object of 1,000,000 columns, where even column names are not materialized either as well as the column data. Even `STRING_ELT` can allocate if the vector it is passed is an ALTREP.